book/src/counter.md

```diff
@@ -0,0 +1,331 @@
+
+```rust
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1get(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong
+) -> jint {
+    let boxed_counter = unsafe { Box::from_raw(ptr as *mut Counter) };
+    let counter_value = boxed_counter.get();
+    Box::into_raw(boxed_counter);
+    return counter_value
+}
+```
+
+## Implement Destroy
+
+To implement destroy, convert the pointer to a `Box`, and let it go out of
+scope. The memory is freed by the `Drop` implementation of `Box`.
+
+```rust
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1destroy(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong
+) {
+    unsafe { Box::from_raw(ptr as *mut Counter) };
+```
+
+## Full Rust Implementation
+
+Putting it all together, the full implementation Counter looks like:
+
+```rust
+use jni::objects::JClass;
+use jni::sys::{jint, jlong};
+use jni::JNIEnv;
+
+struct Counter {
+    count: i32,
+}
+
+impl Counter {
+    fn new() -> Self {
+        Counter { count: 0 }
+    }
+
+    fn increment(&mut self) -> i32 {
+        self.count += 1;
+        self.count
+    }
+
+    fn get(&self) -> i32 {
+        self.count
+    }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1new(
+    _env: JNIEnv,
+    _class: JClass,
+) -> jlong {
+    let boxed_counter = Box::new(Counter::new());
+    Box::into_raw(boxed_counter) as jlong
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1increment(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong,
+) -> jint {
+    let mut boxed_counter = unsafe { Box::from_raw(ptr as *mut Counter) };
+    let counter_value = boxed_counter.increment();
+    Box::into_raw(boxed_counter);
+    return counter_value;
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1get(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong,
+) -> jint {
+    let boxed_counter = unsafe { Box::from_raw(ptr as *mut Counter) };
+    let counter_value = boxed_counter.get();
+    Box::into_raw(boxed_counter);
+    return counter_value;
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1destroy(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong,
+) {
+    unsafe { Box::from_raw(ptr as *mut Counter) };
+}
+```
+
+## Summary
+
+In this section, we've covered how to let a heap-allocated Rust object outlive
+its stack for the purpose of cross-language memory management. We will return to
+this API in the [Cleaning](./cleaning.md) section, where we will see how we can
+make it harder to run into memory leaks, and in [Java Objects in
+Rust](./java_objects_in_rust.md), where we will discuss using arrays of `Counter`.
+
+## Exercises
+
+1. What happens if destroy is never called, or gets called multiple times?
+2. Does the memory created by Rust count towards heap size limits set through
```

Comment:
The ultimate answer from async-profiler author, Andrei Pangin: https://www.youtube.com/watch?v=c755fFv1Rnk :-)